lib/net/x86-64-linux/Net.v3

```diff
@@ -211,37 +378,74 @@ class ClientSocket extends Socket {
 		if (ret == -1 || sockNameRet == -1) {
 			return -1;
 		}
+		setBound(true);
+		setConnected(true);
 		// XXX: Why can't we set `def var` fields from a subclass?
 		setLocalPort((int.view(retSockAddr[2]) << 8) | int.view(retSockAddr[3]));
 		return ret;
 	}
 }
 
-// A socket that accepts connections from incoming peers.
+// A socket that accepts connections from incoming peers. This can only be used
+// as a TCP socket.
 class ServerSocket extends Socket {
-	// Associated sockaddr structure. Is instantiated in bind()
+	// Associated sockaddr structure. Is instantiated in constructor
 	private var sockAddr: Array<byte>;
 
-	new(addr: InetAddress, localPort: int) super(addr, localPort, -1 /* port */, -1 /* fd */, true) { }
+	new(addr: InetAddress, localPort: int) super(addr, -1 /* port */, localPort, SocketType.Tcp, -1 /* fd */, true) {
+		sockAddr = toSockAddr(addr, localPort);
+		var ret = bind();
+		if (ret == -1) {
+			System.error("Error", "Could not bind socket!");
```

Comment:
`System.error` brings down the process, which is pretty violent. These error conditions should be encoded as return values to allow the surrounding program to do something. Thus, this functionality shouldn't really be in the constructor. Same for base behavior. Constructors should be to initialize the object, and possibly-failing operations need to be separate.

I haven't thought about this deeply for networking, but one way would be move some of the functionality into component methods (e.g. helpers that create these objects (or `null` if failure). I'm not great at API design, so take that with a grain of salt.

Reply:
Admittedly `System.error` is quite spartan. I was trying to make the library a bit more hands-free wherein you don't have to manually `bind()` then `listen()` etc etc. I'm thinking maybe a new method like `open()` (or `start()`?) that will perform all the actions.

With regard to errors (in a different comment, but related to this), I can make something like a result type (`type NetResult<T> { Result(v: T); Error(e: NetError); }`; where `NetError` is some `enum` of possible errors). Does that sound better?

Reply:
Yeah, that sounds alright for now. At some point when more libraries accumulate we'll know better what looks good.

Reply:
Should the call to `Net.socket()` in the `Socket` constructor also be moved? It sets the `fd` field, so I'm of the opinion it belongs in the constructor. But it is a syscall and it can fail. I'm leaning towards keeping it in the `Socket` constructor unless you have an opinion on this.

Reply:
Just give drop a comment here.

Although it feels really weird from an API perspective, my two cents is that it shouldn't be in the constructor, because there's no way to recover / indicate failure to the caller from a constructor. Maybe if there was a better error handling story you could hook into that instead, but right now the only way to inform the caller of errors is through a return value.

That said, it's quite unlikely this would fail...

Reply:
Yeah going by the error return values for the `socket()` syscall, it is unlikely to fail for the (hardcoded) values we have. But I wanted to check if it is considered okay to use it in the constructor, or maybe just add it to the `open()` method that I'll make.